lightbulb/converters.py

```diff
@@ -463,64 +464,80 @@ async def foo(ctx, foo: Greedy[int]):
                 ...
     """
 
-    pass
 
+class _ConverterT(typing.Protocol[T_co]):
+    def __call__(self, arg: WrappedArg) -> typing.Union[T_co, typing.Coroutine[None, None, T_co]]:
+        ...
 
-_converter_T = typing.Union[
-    typing.Callable[[WrappedArg], T], typing.Callable[[WrappedArg], typing.Coroutine[None, None, T]]
-]
+
+class _BaseConverter(typing.Protocol[T_co]):
+    async def convert(self, context: context_.Context, arg_string: str, *, parse: bool) -> typing.Tuple[T_co, str]:
+        ...
 
 
 class _Converter:
     __slots__ = ("conversion_func",)
 
-    def __init__(self, conversion_func: _converter_T) -> None:
+    def __init__(self, conversion_func: _ConverterT) -> None:
         self.conversion_func = conversion_func
 
-    async def convert(self, context: context_.Context, arg_string: str, *, parse=True) -> typing.Tuple[T, str]:
+    async def convert(self, context: context_.Context, arg_string: str, *, parse: bool = True) -> typing.Tuple[T, str]:
         args, remainder = arg_string, ""
         if parse:
             sv = stringview.StringView(arg_string)
-            args, remainder = sv.deconstruct_str(max_parse=1)
-            args = " ".join(args)
+            parsed, remainder = sv.deconstruct_str(max_parse=1)
+            args = " ".join(parsed)
 
-        converted_arg = await utils.maybe_await(self.conversion_func, WrappedArg(args, context))
+        arguments = []
+        if self.conversion_func is not None.__class__:
+            arguments.append(WrappedArg(args, context))
+        else:
+            remainder = args
+
+        converted_arg = await utils.maybe_await(self.conversion_func, *arguments)
         return converted_arg, remainder
 
 
 class _UnionConverter:
-    __slots__ = ("converters",)
+    __slots__ = ("converters", "has_none_type")
 
-    def __init__(self, *converters: _Converter) -> None:
+    def __init__(self, *converters: _BaseConverter[T], has_none_type: bool) -> None:
         self.converters = converters
+        self.has_none_type = has_none_type
```

Comment:
I think you have overcomplicated this. I would leave the converter as it was as it seems to work, and instead just wrap it in a defaulting converter if NoneType is present when the converters are resolved